File: bidi-stream/NovaS2SWebSocketExample/app/src/main/java/org/example/utility/OutputEventsInteractObserver.java
package org.example.utility;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.eclipse.jetty.websocket.api.Session;
import org.json.JSONArray;
import org.json.JSONObject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicReference;


class ChatTurn {
    private final String role;
    private final String content;

    public ChatTurn(String role, String content) {
        this.role = role;
        this.content = content;
    }

    public String getRole() {
        return role;
    }

    public String getContent() {
        return content;
    }
}

public class OutputEventsInteractObserver implements InteractObserver<String> {
    private static final Logger log = LoggerFactory.getLogger(OutputEventsInteractObserver.class);
    private final ObjectMapper objectMapper;
    private final Session session;
    private final List<ChatTurn> localChatHistory;
    private String promptId = "";
    private NovaReasoningModel novaReasoningModel;
    AtomicReference<String> toolUseId = new AtomicReference<>("");
    AtomicReference<String> toolUseContent = new AtomicReference<>("");

    public OutputEventsInteractObserver(Session session, NovaReasoningModel novaReasoningModel) {
        this.session = session;
        this.objectMapper = new ObjectMapper();
        this.localChatHistory = new ArrayList<>();
        this.novaReasoningModel = novaReasoningModel;
    }

    @Override
    public void onNext(String msg) {
        try {
            boolean shouldSendMsgToUI = true;
            JsonNode rootNode = objectMapper.readTree(msg);
            JsonNode eventNode = rootNode.get("event");

            if (eventNode != null) {
                if (eventNode.has("completionStart")) {
                    handleCompletionStart(eventNode.get("completionStart"));
                } else if (eventNode.has("contentStart")) {
                    if (eventNode.get("contentStart").get("type") == null) {
                        shouldSendMsgToUI = false;
                    }
                    handleContentStart(eventNode.get("contentStart"));
                } else if (eventNode.has("textOutput")) {
                    if (!eventNode.get("textOutput").get("content").asText().startsWith("Speculative")) {
                        localChatHistory.add(new ChatTurn(eventNode.get("textOutput").get("role").asText(), eventNode.get("textOutput").get("content").asText()));
                    }
                    handleTextOutput(eventNode.get("textOutput"));
                } else if (eventNode.has("audioOutput")) {
                    handleAudioOutput(eventNode.get("audioOutput"));
                } else if (eventNode.has("toolUse")) {
                    shouldSendMsgToUI = false;
                    toolUseId.set(eventNode.get("toolUse").get("toolUseId").asText());
                    toolUseContent.set(eventNode.get("toolUse").get("content").asText());

                } else if (eventNode.has("contentEnd")) {
                    if ("TOOL".equals(eventNode.get("contentEnd").get("type").asText())) {
                        handleToolUse(eventNode);
                        shouldSendMsgToUI = false;
                    }
                    handleContentEnd(eventNode.get("contentEnd"));
                } else if (eventNode.has("completionEnd")) {
                    handleCompletionEnd(eventNode.get("completionEnd"));
                }
            }

            if (shouldSendMsgToUI) {
                sendToUI(msg);
            }

        } catch (Exception e) {
            log.error("Error processing message", e);
            onError(e);
        }
    }

    private void handleCompletionStart(JsonNode node) {
        log.info("Completion started for node: {}", node);
        promptId = node.get("promptName").asText();
        log.info("Completion started with promptId: {}", promptId);
    }

    private void handleContentStart(JsonNode node) {
        log.info("Content started for node: {}", node);
        String contentId = node.get("contentId").asText();
        log.info("Content started with contentId: {}", contentId);
    }

    private void handleTextOutput(JsonNode node) {
        log.info("Text output for node: {}", node);
        String content = node.get("content").asText();
        String role = node.get("role").asText();
        log.info("Received text output: {} from {}", content, role);
    }

    private void handleAudioOutput(JsonNode node) {
        log.info("Audio output for node: {}", node);
        String content = node.get("content").asText();
        String role = node.get("role").asText();
        log.info("Received audio output {} from {}", content, role);
    }

    private void handleToolUse(JsonNode node) {
        log.info("ToolUse for node: {}", node);
        String localChatHistory = createChatHistoryFromLocal();
        log.info("Actual chat History: {}", toolUseContent.get());
        log.info("Local Chat History: {}", localChatHistory);

        novaReasoningModel.processRequest(NovaReasoningModel.createContext(promptId, toolUseId.get()), toolUseContent.get());
        log.info("Tool use: {} with content: {}", toolUseId.get(), toolUseContent.get());
    }

    private void handleContentEnd(JsonNode node) {
        log.info("Content end for node: {}", node);
        String contentId = node.get("contentId").asText();
        String stopReason = node.has("stopReason") ? node.get("stopReason").asText() : "";
        log.info("Content ended: {} with reason: {}", contentId, stopReason);
    }

    private void handleCompletionEnd(JsonNode node) {
        log.info("Completion end for node: {}", node);
        String stopReason = node.has("stopReason") ? node.get("stopReason").asText() : "";
        log.info("Completion ended with reason: {}", stopReason);
    }

    private void sendToUI(String msg) {
        try {
            if (session.isOpen()) {
                session.getRemote().sendString(msg);
            } else {
                log.debug("Ignoring as session is already closed");
            }
        } catch (Exception e) {
            log.error("Error sending message to UI", e);
        }
    }

    private String createChatHistoryFromLocal() {
        JSONArray messageJsonArray = new JSONArray();
        for (ChatTurn cc : localChatHistory) {
            JSONObject messageObj = new JSONObject();
            messageObj.put("role", cc.getRole());
            messageObj.put("content", cc.getContent());
            messageJsonArray.put(messageObj);
        }
        JSONObject chatHistoryObj = new JSONObject();
        chatHistoryObj.put("chatHistory", messageJsonArray);
        return chatHistoryObj.toString();
    }

    @Override
    public void onComplete() {
        log.info("Output complete");
        try {
            session.close(1000, "Output complete");
        } catch (Exception e) {
            log.error("Error closing session", e);
        }
    }

    @Override
    public void onError(Exception error) {
        log.error("Error occurred", error);
        try {
            session.close(1011, "Error occurred: " + error.getMessage());
        } catch (Exception e) {
            log.error("Error closing session", e);
        }
    }
}

File: speech-to-speech/amazon-nova-2-sonic/sample-codes/websocket-dotnet/Utility/OutputEventsInteractObserver.cs
using System.Net.WebSockets;
using System.Text;
using Microsoft.Extensions.Logging;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace NovaSonicWebSocket.Utility;

public class ChatTurn
{
    public string Role { get; }
    public string Content { get; }

    public ChatTurn(string role, string content)
    {
        Role = role;
        Content = content;
    }
}

public class OutputEventsInteractObserver : IInteractObserver<string>
{
    private readonly WebSocket _webSocket;
    private readonly ILogger _logger;
    private readonly List<ChatTurn> _localChatHistory = new();
    private readonly AtomicReference<string> _promptId;
    private readonly AtomicReference<string> _toolUseId = new(string.Empty);
    private readonly AtomicReference<string> _toolUseContent = new(string.Empty);
    private readonly AtomicReference<string> _toolName = new(string.Empty);
    private readonly AtomicReference<string> _role = new(string.Empty);
    private readonly AtomicReference<string> _generationStage = new(string.Empty);
    private readonly ToolResultProcessor _toolResultProcessor;
    private IInteractObserver<string>? _inputObserver;

    public OutputEventsInteractObserver(WebSocket webSocket, ILogger logger, AtomicReference<string> promptId)
    {
        _webSocket = webSocket;
        _logger = logger;
        _promptId = promptId;
        _toolResultProcessor = new ToolResultProcessor(logger);
    }

    public void SetInputObserver(IInteractObserver<string>? inputObserver)
    {
        _inputObserver = inputObserver;
    }

    public void OnNext(string msg)
    {
        try
        {
            var shouldSendMsgToUI = true;
            var jsonObject = JObject.Parse(msg);
            var eventNode = jsonObject["event"];

            if (eventNode != null)
            {
                if (eventNode["completionStart"] != null)
                {
                    HandleCompletionStart(eventNode["completionStart"]);
                }
                else if (eventNode["contentStart"] != null)
                {
                    if (eventNode["contentStart"]?["type"] == null)
                    {
                        shouldSendMsgToUI = false;
                    }
                    HandleContentStart(eventNode["contentStart"]);
                }
                else if (eventNode["textOutput"] != null)
                {
                    if (_generationStage.Value != "SPECULATIVE")
                    {
                        var content = eventNode["textOutput"]?["content"]?.ToString() ?? string.Empty;
                        _localChatHistory.Add(new ChatTurn(_role.Value, content));
                    }
                    HandleTextOutput(eventNode["textOutput"]);
                }
                else if (eventNode["audioOutput"] != null)
                {
                    HandleAudioOutput(eventNode["audioOutput"]);
                }
                else if (eventNode["toolUse"] != null)
                {
                    shouldSendMsgToUI = false;
                    _toolUseId.Value = eventNode["toolUse"]?["toolUseId"]?.ToString() ?? string.Empty;
                    _toolUseContent.Value = eventNode["toolUse"]?["content"]?.ToString() ?? string.Empty;
                    _toolName.Value = eventNode["toolUse"]?["toolName"]?.ToString() ?? string.Empty;
                    _logger.LogInformation("tool use: {ToolUse}", eventNode["toolUse"]);
                }
                else if (eventNode["contentEnd"] != null)
                {
                    if (eventNode["contentEnd"]?["type"]?.ToString() == "TOOL")
                    {
                        HandleToolUse(eventNode);
                        shouldSendMsgToUI = false;
                    }
                    HandleContentEnd(eventNode["contentEnd"]);
                }
                else if (eventNode["completionEnd"] != null)
                {
                    HandleCompletionEnd(eventNode["completionEnd"]);
                }
                else if (eventNode["sessionEnd"] != null)
                {
                    HandleSessionEnd();
                }
                else if (eventNode["usageEvent"] != null)
                {
                    _logger.LogDebug("Parsing usage metrics jsonText={UsageEvent}", eventNode);
                }
            }

            if (shouldSendMsgToUI)
            {
                // Fire and forget - don't block the observer thread
                _ = SendMessageAsync(msg);
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error processing message");
            OnError(ex);
        }
    }

    public void OnComplete()
    {
        _logger.LogInformation("Stream completed");
        _inputObserver?.OnComplete();
    }

    public void OnError(Exception error)
    {
        _logger.LogError(error, "Error in output observer");

        try
        {
            var errorMessage = new
            {
                error = new
                {
                    message = error.Message,
                    type = error.GetType().Name
                }
            };
            _logger.LogError("Sending error message to client: {ErrorMessage}", JsonConvert.SerializeObject(errorMessage));
            // Fire and forget - don't block on error handling
            _ = SendMessageAsync(JsonConvert.SerializeObject(errorMessage));
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error sending error message to client");
        }
    }

    private void HandleCompletionStart(JToken? node)
    {
        if (node != null)
        {
            _promptId.Value = node["promptName"]?.ToString() ?? string.Empty;
            _logger.LogInformation("Completion started with promptId: {PromptId}", _promptId.Value);
            _logger.LogInformation("Nova Sonic session: {SessionId}", node["sessionId"]?.ToString() ?? string.Empty);
        }
    }

    private void HandleContentStart(JToken? node)
    {
        if (node != null)
        {
            _logger.LogInformation("Content started for node: {Node}", node);
            try
            {
                if (node["additionalModelFields"] != null)
                {
                    var additionalModelFieldsStr = node["additionalModelFields"]?.ToString() ?? string.Empty;
                    var additionalFields = JObject.Parse(additionalModelFieldsStr);
                    // FINAL is the option for USER
                    // SPECULATIVE is the option before audio generation for ASSISTANT
                    // FINAL is the other option after audio generation for ASSISTANT
                    _generationStage.Value = additionalFields["generationStage"]?.ToString() ?? string.Empty;
                }
                if (node["role"] != null)
                {
                    // USER, ASSISTANT, or TOOL
                    _role.Value = node["role"]?.ToString() ?? string.Empty;
                }
            }
            catch (Exception e)
            {
                _logger.LogError(e, "Error processing content event");
            }

            var contentId = node["contentId"]?.ToString() ?? string.Empty;
            _logger.LogInformation("Content started with contentId: {ContentId}", contentId);
        }
    }

    private void HandleTextOutput(JToken? node)
    {
        if (node != null)
        {
            var content = node["content"]?.ToString() ?? string.Empty;
            _logger.LogInformation("Received text output: {Content} from {Role}", content, _role.Value);
        }
    }

    private void HandleAudioOutput(JToken? node)
    {
        if (node != null)
        {
            var content = node["content"]?.ToString() ?? string.Empty;
            _logger.LogDebug("Received audio output {Content} from {Role}", content, _role.Value);
        }
    }

    private void HandleToolUse(JToken? node)
    {
        if (node == null)
        {
            _logger.LogWarning("Received null node in handleToolUse");
            return;
        }

        try
        {
            ValidateToolUseParameters();
            ProcessToolUseAsync();
        }
        catch (InvalidOperationException e)
        {
            _logger.LogError(e, "Tool use processing failed: {Message}", e.Message);
        }
    }

    private void ValidateToolUseParameters()
    {
        if (string.IsNullOrEmpty(_toolName.Value) || 
            string.IsNullOrEmpty(_toolUseId.Value) || 
            string.IsNullOrEmpty(_promptId.Value))
        {
            throw new InvalidOperationException("Missing required tool use parameters");
        }
    }

    private void ProcessToolUseAsync()
    {
        _logger.LogDebug("Processing tool use asynchronously");
        var localChatHistory = CreateChatHistoryFromLocal();
        LogChatHistories(localChatHistory);

        // Process the tool use asynchronously
        var task = _toolResultProcessor.ProcessToolUseAsync(
            _promptId.Value, 
            _toolUseId.Value, 
            _toolName.Value, 
            _toolUseContent.Value);

        task.ContinueWith(t =>
        {
            if (t.IsCompletedSuccessfully && t.Result != null)
            {
                var response = t.Result;
                SendStart(response.ContentId, response.ToolUseId);
                SendToolResult(response);
                SendEnd(response.ContentId);
            }
            else if (t.IsFaulted)
            {
                _logger.LogError(t.Exception, "Error processing tool result");
                // Send error response to the model
                var errorContentId = Guid.NewGuid().ToString();
                SendStart(errorContentId, _toolUseId.Value);
                SendErrorToolResult(errorContentId, t.Exception);
                SendEnd(errorContentId);
            }
        });
    }

    private void SendToolResult(ToolResultProcessor.ToolResultResponse response)
    {
        if (_inputObserver != null)
        {
            try
            {
                // Create the "toolResult" object
                var toolResultNode = new JObject
                {
                    ["promptName"] = response.PromptId,
                    ["contentName"] = response.ContentId,
                    ["content"] = response.Content // Content is already properly formatted JSON string
                };

                // Create the final JSON structure
                var eventNode = new JObject
                {
                    ["toolResult"] = toolResultNode
                };

                var rootNode = new JObject
                {
                    ["event"] = eventNode
                };

                var jsonPayload = rootNode.ToString(Formatting.None);
                _logger.LogDebug("Tool Result - Tool Result payload: {Payload}", jsonPayload);
                _inputObserver.OnNext(jsonPayload);
            }
            catch (Exception e)
            {
                throw new Exception("Error creating JSON payload for toolResult", e);
            }
        }
    }

    private void SendErrorToolResult(string contentId, Exception? ex)
    {
        if (_inputObserver != null)
        {
            try
            {
                var errorContent = new JObject
                {
                    ["error"] = $"Tool execution failed: {ex?.Message}"
                };

                var toolResultNode = new JObject
                {
                    ["promptName"] = _promptId.Value,
                    ["contentName"] = contentId,
                    ["content"] = errorContent.ToString(Formatting.None)
                };

                // Create the final JSON structure
                var eventNode = new JObject
                {
                    ["toolResult"] = toolResultNode
                };

                var rootNode = new JObject
                {
                    ["event"] = eventNode
                };

                var jsonPayload = rootNode.ToString(Formatting.None);
                _logger.LogDebug("Sending error tool result: {Payload}", jsonPayload);
                _inputObserver.OnNext(jsonPayload);
            }
            catch (Exception e)
            {
                _logger.LogError(e, "Failed to send error tool result");
            }
        }
    }

    private void SendStart(string contentId, string toolUseId)
    {
        if (_inputObserver != null)
        {
            try
            {
                var textInputConfigNode = new JObject
                {
                    ["mediaType"] = "text/plain"
                };

                var toolResultInputConfigNode = new JObject
                {
                    ["toolUseId"] = toolUseId,
                    ["type"] = "TEXT",
                    ["textInputConfiguration"] = textInputConfigNode
                };

                var contentStartNode = new JObject
                {
                    ["promptName"] = _promptId.Value,
                    ["contentName"] = contentId,
                    ["interactive"] = false,
                    ["type"] = "TOOL",
                    ["role"] = "TOOL",
                    ["toolResultInputConfiguration"] = toolResultInputConfigNode
                };

                var eventNode = new JObject
                {
                    ["contentStart"] = contentStartNode
                };

                var rootNode = new JObject
                {
                    ["event"] = eventNode
                };

                var contentStart = rootNode.ToString(Formatting.None);
                _logger.LogDebug("Tool Result - Content Start payload: {Payload}", contentStart);
                _inputObserver.OnNext(contentStart);
            }
            catch (Exception e)
            {
                throw new Exception("Error creating JSON payload for Tool Result contentStart", e);
            }
        }
    }

    private void SendEnd(string contentId)
    {
        if (_inputObserver != null)
        {
            try
            {
                var contentEndNode = new JObject
                {
                    ["promptName"] = _promptId.Value,
                    ["contentName"] = contentId
                };

                var eventNode = new JObject
                {
                    ["contentEnd"] = contentEndNode
                };

                var rootNode = new JObject
                {
                    ["event"] = eventNode
                };

                var contentEnd = rootNode.ToString(Formatting.None);
                _logger.LogDebug("Tool Result - Content End payload: {Payload}", contentEnd);
                _inputObserver.OnNext(contentEnd);
            }
            catch (Exception e)
            {
                throw new Exception("Error creating JSON payload for Tool Result contentEnd", e);
            }
        }
    }

    private void LogChatHistories(string localChatHistory)
    {
        _logger.LogDebug("Actual chat History: {ActualHistory}", _toolUseContent.Value);
        _logger.LogDebug("Local Chat History: {LocalHistory}", localChatHistory);
    }

    private void HandleContentEnd(JToken? node)
    {
        if (node != null)
        {
            var contentId = node["contentId"]?.ToString() ?? string.Empty;
            var stopReason = node["stopReason"]?.ToString() ?? string.Empty;
            _logger.LogInformation("Content ended: {ContentId} with reason: {StopReason}", contentId, stopReason);
        }
    }

    private void HandleCompletionEnd(JToken? node)
    {
        if (node != null)
        {
            var stopReason = node["stopReason"]?.ToString() ?? string.Empty;
            _logger.LogInformation("Completion ended with reason: {StopReason}", stopReason);
        }
    }

    private void HandleSessionEnd()
    {
        _logger.LogInformation("Session ended");
        _inputObserver?.OnComplete();
    }

    private string CreateChatHistoryFromLocal()
    {
        var messageJsonArray = new JArray();
        foreach (var cc in _localChatHistory)
        {
            var messageObj = new JObject
            {
                ["role"] = cc.Role,
                ["content"] = cc.Content
            };
            messageJsonArray.Add(messageObj);
        }

        var chatHistoryObj = new JObject
        {
            ["chatHistory"] = messageJsonArray
        };

        return chatHistoryObj.ToString(Formatting.None);
    }

    private async Task SendMessageAsync(string message)
    {
        if (_webSocket.State == WebSocketState.Open)
        {
            var buffer = Encoding.UTF8.GetBytes(message);
            await _webSocket.SendAsync(
                new ArraySegment<byte>(buffer),
                WebSocketMessageType.Text,
                true,
                CancellationToken.None).ConfigureAwait(false);
        }
    }
}
